File: app/src/main/java/de/openfiresource/openpager/models/UserRepository.java
package de.openfiresource.openpager.models;

import com.google.firebase.iid.FirebaseInstanceId;

import java.net.UnknownHostException;

import javax.inject.Inject;
import javax.inject.Singleton;

import de.openfiresource.openpager.models.api.Device;
import de.openfiresource.openpager.models.api.UserLogin;
import de.openfiresource.openpager.utils.DeviceInfoHelper;
import de.openfiresource.openpager.utils.Preferences;
import io.reactivex.Completable;

@Singleton
public class UserRepository {

    private final Rest rest;

    private final Preferences preferences;

    @Inject
    UserRepository(Rest rest, Preferences preferences) {
        this.rest = rest;
        this.preferences = preferences;
    }

    public Completable login(String email, String password) {
        UserLogin user = new UserLogin.Builder()
                .email(email)
                .password(password)
                .build();

        return rest.login(user)
                .doOnSuccess(userKey -> {
                    if (userKey.getKey() != null) {
                        preferences.getUserKey().set(userKey.getKey());
                    }
                })
                .flatMapCompletable(userKey -> sendDeviceInfo(FirebaseInstanceId.getInstance().getToken()))
                .doOnError(throwable -> preferences.getUserKey().set(""));
    }

    public Completable logout() {
        return rest.logout()
                .onErrorResumeNext(throwable -> {
                    if (throwable instanceof UnknownHostException) {
                        return Completable.error(throwable);
                    }

                    // We don't know if it's a server error so we have to log out the user
                    preferences.getUserKey().set("");
                    return Completable.complete();
                })
                .doOnComplete(() -> preferences.getUserKey().set(""));
    }

    public Completable sendDeviceInfo(String fcmToken) {
        Device device = DeviceInfoHelper.create(fcmToken);
        return rest.putDeviceInfo(device);
    }
}

File: app/src/main/java/de/openfiresource/openpager/models/UserRepository.kt
package de.openfiresource.openpager.models

import com.google.firebase.iid.FirebaseInstanceId
import de.openfiresource.openpager.models.api.UserLogin
import de.openfiresource.openpager.utils.DeviceInfoHelper
import de.openfiresource.openpager.utils.Preferences
import io.reactivex.Completable
import java.net.UnknownHostException
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class UserRepository @Inject internal constructor(private val rest: Rest, private val preferences: Preferences) {

    fun login(email: String, password: String): Completable {
        val user = UserLogin(email, password)

        return rest.login(user)
                .doOnSuccess { userKey ->
                    userKey.key?.let { preferences.userKey.set(it) }
                }
                .flatMapCompletable { (_) -> sendDeviceInfo(FirebaseInstanceId.getInstance().token) }
                .doOnError { preferences.userKey.set("") }
    }

    fun logout(): Completable {
        return rest.logout()
                .onErrorResumeNext { throwable ->
                    if (throwable is UnknownHostException) {
                        return@onErrorResumeNext Completable.error(throwable)
                    }

                    // We don't know if it's a server error so we have to log out the user
                    preferences.userKey.set("")
                    Completable.complete()
                }
                .doOnComplete { preferences.userKey.set("") }
    }

    fun sendDeviceInfo(fcmToken: String?): Completable {
        fcmToken?.let {
            val device = DeviceInfoHelper.create(it)
            return rest.putDeviceInfo(device)
        }
        return Completable.error(Throwable("No FCM-Token delivered"));
    }
}
